plugins/org.eclipse.epsilon.eol.engine/src/org/eclipse/epsilon/eol/ImportManager.java

```diff
@@ -0,0 +1,109 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import org.eclipse.epsilon.common.module.IModule;
+import org.eclipse.epsilon.common.util.UriUtil;
+import org.eclipse.epsilon.eol.dom.Import;
+
+public class ImportManager implements IImportManager {
+
+	/**
+	 * Note: this cache key scheme assumes that adapt() works the same for
+	 * all instances of a given IEolModule implementation (so it does not
+	 * matter which specific instance is the parent of an imported module,
+	 * so long as it is the same type).
+	 *
+	 * In other words, it assumes that adapt() only works from its arguments,
+	 * and that it does not use any state of the IEolModule instance in question.
+	 */
+	protected static class ImportCacheKey {
+		public final URI uri;
+		public final Class<? extends IModule> parentModuleClass;
+
+		public ImportCacheKey(Class<? extends IModule> parentModuleClass, URI uri) {
+			this.uri = uri;
+			this.parentModuleClass = parentModuleClass;
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hash(parentModuleClass, uri);
+		}
+
+		@Override
+		public boolean equals(Object obj) {
+			if (this == obj)
+				return true;
+			if (obj == null)
+				return false;
+			if (getClass() != obj.getClass())
+				return false;
+
+			ImportCacheKey other = (ImportCacheKey) obj;
+			return Objects.equals(parentModuleClass, other.parentModuleClass) && Objects.equals(uri, other.uri);
+		}
+	}
+
+	protected Map<ImportCacheKey, IModule> cache = new HashMap<>();
+
+	@Override
+	public void loadModuleForImport(Import import_, Class<? extends IModule> moduleImplClass, URI baseURI) throws URISyntaxException {
+		String importPath = import_.getPath();
+		
+		URI importUri = UriUtil.resolve(importPath, baseURI).normalize();
+
+		final IEolModule parentModule = import_.getParentModule();
+		final ImportCacheKey cacheKey = new ImportCacheKey(parentModule.getClass(), importUri);
+
+		IModule module = cache.get(cacheKey);
+		if (module != null) {
+			if (!moduleImplClass.isInstance(module)) {
+				String message = String.format("Module %s is not an instance of %s. Instead it is %s.",
+						importUri.toString(), moduleImplClass.getName(), module.getClass().getName());
+				throw new IllegalArgumentException(message);
+			}
+
+			import_.setImportedModule(module);
+			import_.setLoaded(true);
+		} else {
+			try {
+				module = moduleImplClass.getDeclaredConstructor().newInstance();
+			} catch (Exception e) {
+				e.printStackTrace();
+				return;
+			}
+
+			/*
+			 * The module must be added to the cache before import_.load is called,
+			 * otherwise there will be infinite recursion in the case of circular import.
+			 */
+			cache.put(cacheKey, module);
+
+			/*
+			 * If possible, ensure child modules share this import manager so the cache is
+			 * most effective.
+			 */
+			if (module instanceof IEolModule) {
+				IEolModule eolModule = (IEolModule) module;
+				eolModule.setImportManager(this);
```

Comment:
I think it may make more sense to just have `getImportManager()` delegate to the parent module's `getImportManager()` method whenever it is set, instead of needing this explicit method call.

Reply:
Nevermind - this option is better, as we cannot be sure what the `getImportManager()` method of an arbitrary `IEolModule` implementation will do. It could not honor this idea of using the import manager from the module, by mistake.